alchemist/alchemist-cognitive-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/actions/AbstractOrienting.kt

```diff
@@ -0,0 +1,417 @@
+        val newRoom = environment.graph().nodeContaining(currentPosition)
+            ?: throw IllegalStateException("pedestrian is not inside a room")
+        pedestrian.registerVisit(newRoom)
+        when {
+            ::nextRoom.isInitialized -> {
+                if (nextRoom.contains(currentPosition)) {
+                    currRoom = nextRoom
+                }
+                /*
+                 * We should have reached nextRoom, but we are in a different room (newRoom),
+                 * in such case we return back to the previous room (stored in currRoom), so
+                 * as to correctly follow our route.
+                 * In the future, something more sophisticated could be done (e.g. recomputing
+                 * the route).
+                 */
+                else {
+                    val prevRoom = currRoom
+                    currRoom = newRoom
+                    val doorsLeadingBack = environment.graph().outgoingEdgesOf(currRoom)
+                        .filter { environment.graph().getEdgeTarget(it) == prevRoom }
+                    moveToClosestDoor(doors = doorsLeadingBack)
+                    return
+                }
+            }
+            else -> currRoom = newRoom
+        }
+        environment.graph().destinationsWithin(currRoom).let { destinationsInSight ->
+            if (destinationsInSight.isNotEmpty()) {
+                route.clear()
+                subdestination = destinationsInSight.first()
+                state = State.MOVING_TO_FINAL
+                return
+            }
+        }
+        /*
+         * If a sub-destination of the route is in sight, updates the route removing
+         * all the sub-destination up to the one encountered.
+         */
+        if (route.isNotEmpty() && route.any { currRoom.contains(it.centroid) }) {
+            for (i in 0..route.indexOfLast { currRoom.contains(it.centroid) }) {
+                route.removeAt(0)
+            }
+        }
+        val rankings = if (route.isNotEmpty()) {
+            computeEdgeRankings(currRoom, route[0].centroid)
+        } else null
+        /*
+         * The pedestrian can see all the edges outgoing from the current room.
+         */
+        val minEdge = environment.graph().outgoingEdgesOf(currRoom)
+            .minWith(
+                compareBy({
+                    weight(it, rankings?.get(it))
+                }, {
+                    /*
+                     * nearest door heuristic
+                     */
+                    crossingPoint(it).distanceTo(currentPosition)
+                })
+            )
+        if (minEdge != null) {
+            nextRoom = environment.graph().getEdgeTarget(minEdge)
+            subdestination = crossingPoint(minEdge)
+            targetDoor = minEdge
+            state = State.MOVING_TO_DOOR
+        }
+        /*
+         * Closed room, we can't move anywhere.
+         */
+        else {
+            state = State.ARRIVED
+        }
+    }
```

Comment:
Looks to me the goal of this method is to set the new state. As it is, it's very C-ish and not much kotlin-ish. A refactor with a `private fun nextState(...): State` computing the next state and this method reduced to a caller of `nextState` with possibly argument setup would be probably more readable.

Reply:
The goal of `inNewRoom()` is to set the new state and the related variables (such as `subdestination`, `currRoom`, etc) when the agent is in a new room. It's part of the finite state machine describing the orienting behaviour. I agree it's very C-ish, but I doubt whether defining a `private fun nextState(...): State` would make it any cleaner, as that method would be almost a copy-paste of the current one. 
I propose to maintain the policy regarding the setting of state variables within this method, and export anything else in small new methods, as for example `updateRoute()` and `computeRankings()`. It won't make the method more Kotlin-ish, but probably more readable.

Reply:
ok